src/geometry/utility/geometry.cc

```diff
@@ -0,0 +1,308 @@
+    // does not intersect with other parts of the boundary
+    const Segment3d bound_connection(*other_point, *candidate);
+    const auto centerline_point_candidate = MakeCenterpoint(bound_connection.first, bound_connection.second);
+    const Segment3d centerline_candidate{last_point, centerline_point_candidate};
+    if (!bounds.Intersects2d(centerline_candidate)) {
+      distance = candidate_distance;
+      closest_position = candidate;
+    }
+    return false;
+  };
+  if (is_left) {
+    bounds.ForEachPointLeftUntil(other_point,
+                                 non_intersecting_point_loop);  // For each point in the left string to other_point
+  } else {
+    bounds.ForEachPointRightUntil(other_point, non_intersecting_point_loop);
+  }
+  return {closest_position, distance};
+}
+
+}  // namespace
+
+LineString3d ComputeCenterline3d(const LineString3d& left, const LineString3d& right) {
+  LineString3d centerline;
+  BoundChecker bounds(left, right);
+  // Initial point
+  centerline.push_back(MakeCenterpoint(left.first(), right.first()));
+
+  auto left_current = left.begin();
+  auto right_current = right.begin();
+
+  while (left_current != left.end() || right_current != right.end()) {
+    std::optional<double> left_candidate_distance;
+    std::optional<double> right_candidate_distance;
+
+    LineString3d::const_iterator left_candidate;
+    LineString3d::const_iterator right_candidate;
+    // Determine left candidate
+    std::tie(left_candidate, left_candidate_distance) =
+        FindClosestNonintersectingPoint(std::next(left_current), right_current, bounds, centerline.last(), true);
+
+    // Determine right candidate
+    std::tie(right_candidate, right_candidate_distance) =
+        FindClosestNonintersectingPoint(std::next(right_current), left_current, bounds, centerline.last(), false);
+    // Choose the better one
+    if (left_candidate_distance && (!right_candidate_distance || left_candidate_distance <= right_candidate_distance)) {
+      MALIPUT_THROW_UNLESS(left_candidate != left.end());
+
+      const auto& left_point = left[size_t(left_candidate - left.begin())];
+      const auto& right_point = right[size_t(right_current - right.begin())];
+      const auto centerpoint = MakeCenterpoint(left_point, right_point);
+      centerline.push_back(centerpoint);
+      left_current = left_candidate;
+    } else if (right_candidate_distance &&
+               (!left_candidate_distance || left_candidate_distance > right_candidate_distance)) {
+      MALIPUT_THROW_UNLESS(right_candidate != right.end());
+
+      const auto& left_point = left[size_t(left_current - left.begin())];
+      const auto& right_point = right[size_t(right_candidate - right.begin())];
+      const auto centerpoint = MakeCenterpoint(left_point, right_point);
+      centerline.push_back(centerpoint);
+      right_current = right_candidate;
+    } else {
+      // no next point found. We are done here
```

Comment:
Shall we add an info log?

Reply:
You mean for a supposely backtracing? I don't think so. It is part of the behavior breaking here. If added a log here we should add a log for the other cases as it would work as tracing.